hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/ResourceCalculationDriver.java

```diff
@@ -0,0 +1,368 @@
+ * bookkeeper of disposable update information that is used by all children under a common parent.
+ */
+public class ResourceCalculationDriver {
+  protected static final Set<ResourceUnitCapacityType> CALCULATOR_PRECEDENCE =
+      ImmutableSet.of(
+          ResourceUnitCapacityType.ABSOLUTE,
+          ResourceUnitCapacityType.PERCENTAGE,
+          ResourceUnitCapacityType.WEIGHT);
+
+  protected final QueueResourceRoundingStrategy roundingStrategy =
+      new DefaultQueueResourceRoundingStrategy(CALCULATOR_PRECEDENCE);
+  protected final CSQueue parent;
+  protected final QueueCapacityUpdateContext updateContext;
+  protected final Map<ResourceUnitCapacityType, AbstractQueueCapacityCalculator> calculators;
+  protected final Collection<String> definedResources;
+
+  protected final Map<String, ResourceVector> overallRemainingResource = new HashMap<>();
+  protected final Map<String, ResourceVector> batchRemainingResource = new HashMap<>();
+  // Used by ABSOLUTE capacity types
+  protected final Map<String, ResourceVector> normalizedResourceRatio = new HashMap<>();
+  // Used by WEIGHT capacity typet js
+  protected final Map<String, Map<String, Float>> sumWeightsPerLabel = new HashMap<>();
+
+  protected String currentResourceName;
+  protected AbstractQueueCapacityCalculator currentCalculator;
+  protected CSQueue currentChild;
+  protected Map<String, Float> usedResourceByCurrentCalculator = new HashMap<>();
+
+  public ResourceCalculationDriver(
+      CSQueue parent, QueueCapacityUpdateContext updateContext,
+      Map<ResourceUnitCapacityType, AbstractQueueCapacityCalculator> calculators,
+      Collection<String> definedResources) {
+    this.parent = parent;
+    this.updateContext = updateContext;
+    this.calculators = calculators;
+    this.definedResources = definedResources;
+  }
+
+  /**
+   * Returns the parent that is driving the calculation.
+   *
+   * @return a common parent queue
+   */
+  public CSQueue getParent() {
+    return parent;
+  }
+
+  /**
+   * Returns the context that is used throughout the whole update phase.
+   *
+   * @return update context
+   */
+  public QueueCapacityUpdateContext getUpdateContext() {
+    return updateContext;
+  }
+
+  /**
+   * Returns the name of the resource that is currently processed.
+   *
+   * @return resource name
+   */
+  public String getCurrentResourceName() {
+    return currentResourceName;
+  }
+
+  /**
+   * Returns the child that is currently processed.
+   *
+   * @return child queue
+   */
+  public CSQueue getCurrentChild() {
+    return currentChild;
+  }
+
+  /**
+   * Sets the currently evaluated child to a specific queue.
+   *
+   * @param currentChild a child queue
+   */
+  public void setCurrentChild(CSQueue currentChild) {
+    if (currentChild.getParent() != parent) {
+      throw new IllegalArgumentException("Child queue " + currentChild.getQueuePath() + " is not " +
+          "a child of " + parent.getQueuePath());
+    }
```

Comment:
We shouldn't change these independently, the change to child should invoke a change to parent, and parent shouldn't be set independently. Since the calculation requires an ACTUAL queue, just having a parent is not sufficient, so it shouldn't be set independently.